macros/src/types/named.rs

```diff
@@ -88,17 +117,118 @@ fn format_field(
 }
+
+fn extract_option_argument(ty: &Type) -> Result<&Type> {
+    match ty {
+        Type::Path(type_path)
+            if type_path.qself.is_none()
+                && type_path.path.leading_colon.is_none()
+                && type_path.path.segments.len() == 1
+                && type_path.path.segments[0].ident == "Option" =>
+        {
+            let segment = &type_path.path.segments[0];
+            match &segment.arguments {
+                PathArguments::AngleBracketed(args) if args.args.len() == 1 => {
+                    match &args.args[0] {
+                        GenericArgument::Type(inner_ty) => Ok(&inner_ty),
+                        _ => syn_err!("`Option` argument must be a type"),
+                    }
+                }
+                _ => syn_err!("`Option` type must have a single generic argument"),
+            }
+        }
+        _ => syn_err!("`optional` can only be used on an Option<T> type"),
+    }
+}
+
+fn extract_type_args(ty: &Type) -> Option<Vec<&Type>> {
+    let last_segment = match ty {
+        Type::Path(type_path) => type_path.path.segments.last(),
+        _ => None,
+    }?;
+
+    let segment_arguments = match &last_segment.arguments {
+        PathArguments::AngleBracketed(generic_arguments) => {
+            if has_specialized_impl(&last_segment.ident) {
+                None
+            } else {
+                Some(generic_arguments)
+            }
+        }
+        _ => None,
+    }?;
+
+    let type_args: Vec<_> = segment_arguments
+        .args
+        .iter()
+        .filter_map(|arg| match arg {
+            GenericArgument::Type(ty) => Some(ty),
+            _ => None,
+        })
+        .collect();
+    if type_args.is_empty() {
+        return None;
+    }
+
+    Some(type_args)
+}
+
+fn format_type(ty: &Type, generics: &Generics) -> TokenStream {
+    // If the type matches one of the generic parameters, just pass the identifier:
+    if let Some(generic_ident) = generics
+        .params
+        .iter()
+        .filter_map(|param| match param {
+            GenericParam::Type(type_param) => Some(type_param),
+            _ => None,
+        })
+        .find(|type_param| {
+            matches!(
+                ty,
+                Type::Path(type_path)
+                    if type_path.qself.is_none()
+                    && type_path.path.is_ident(&type_param.ident)
+            )
+        })
+        .map(|type_param| type_param.ident.to_string())
+    {
+        return quote!(#generic_ident.to_owned());
+    }
+
+    match extract_type_args(ty) {
+        None => quote!(<#ty as ts_rs::TS>::name()),
+        Some(type_args) => {
+            let args = type_args
+                .iter()
+                .map(|ty| format_type(ty, generics))
+                .collect::<Vec<_>>();
+            let args = quote!(vec![#(#args),*]);
+            quote!(<#ty as ts_rs::TS>::name_with_type_args(#args))
+        }
+    }
+}
+
+fn has_specialized_impl(ident: &Ident) -> bool {
```

Comment:
This one is not as bad as during the first attempt (because arrays and maps are now handled more generically), but I'm still open to suggestions on how to improve this.